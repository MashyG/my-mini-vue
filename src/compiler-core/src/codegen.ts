import { NodeTypes } from './ast'
import {
  CREATE_ELEMENT_VNODE,
  TO_DISPLAY_STRING,
  helperMapName
} from './runtimeHelpers'

export function generate(ast) {
  console.log('ast === ', ast)

  const context = createCodegenContext()
  const { push } = context || {}

  genFunctionPreamble(ast, context)
  push('return ')

  const functionName = 'render'
  const args = ['_ctx', '_cache']
  const signature = args.join(', ')

  push(`function ${functionName}(${signature}) {\n`)
  push(`return `)
  genNode(ast.codegenNode, context)
  push(`\n`)
  push(`}`)

  console.log('context.code ---->>>>> ', context.code)

  return {
    code: context.code
  }
}

function createCodegenContext() {
  const context = {
    code: '',
    push(source) {
      context.code += source
    },
    helper(key) {
      return `_${helperMapName[key]}`
    }
  }
  return context
}

function genFunctionPreamble(ast, context) {
  const { helpers } = ast || {}
  const { length } = helpers || []

  if (length > 0) {
    const VueBinging = 'Vue'
    const aliasHelper = (s) => `${helperMapName[s]}: _${helperMapName[s]}`
    context.push(
      `const { ${helpers.map(aliasHelper).join(', ')} } = ${VueBinging}\n`
    )
  }
}

function genNode(node: any, context) {
  console.log('genNode === ', node)
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, context)
      break
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context)
      break
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context)
      break
    case NodeTypes.ELEMENT:
      genElement(node, context)
      break
    default:
      break
  }
}

// 纯字符串
function genText(node, context) {
  context.push(`'${node.content}'`)
}

// 插值
function genInterpolation(node, context) {
  const { push, helper } = context || {}
  push(`${helper(TO_DISPLAY_STRING)}(`)
  genNode(node.content, context)
  push(`)`)
}

// 表达式
function genExpression(node, context) {
  context.push(`${node.content}`)
}

function genElement(node, context) {
  const { push, helper } = context || {}
  const { tag, children } = node || {}
  console.log('genElement -- children', children)
  push(`${helper(CREATE_ELEMENT_VNODE)}('${tag}', null, `)

  for (let i = 0; i < children.length; i++) {
    const element = children[i]
    genNode(element, context)
  }

  push(')')
}
