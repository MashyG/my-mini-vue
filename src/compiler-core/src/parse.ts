import { NodeTypes } from './ast'

interface ParseContext {
  source: string
}
export interface TagElementType {
  type: NodeTypes
  tag: string
  children: Array<TagElementType>
}

enum TagTypes {
  START,
  END
}

function startsWith(source: string, searchString: string): boolean {
  return source.startsWith(searchString)
}

export function baseParse(content: string): any {
  const context = createParseContext(content)
  return createRoot(parseChildren(context, []))
}

// 根据有限状态机的原理，实现 Parse 流程
function parseChildren(context: ParseContext, ancestor: TagElementType[]) {
  const nodes: any = []

  while (!isEnd(context, ancestor)) {
    let node
    const s = context.source || ''
    if (startsWith(s, '{{')) {
      // 插值
      node = parseInterpolation(context)
    } else if (s[0] === '<') {
      // element
      if (s[1] === '/') {
        // 这里属于 edge case 可以不用关心
        // 处理结束标签
        if (/[a-z]/i.test(s[2])) {
          // 匹配 </div>
          // 需要改变 context.source 的值 -> 也就是需要移动光标
          parseTag(context, TagTypes.END)
          // 结束标签就以为这都已经处理完了，所以就可以跳出本次循环了
          continue
        }
      } else if (/[a-z]/.test(s[1])) {
        node = parseElement(context, ancestor)
      }
    }

    // text 作为默认解析
    if (!node) {
      node = parseText(context)
    }

    nodes.push(node)
  }

  return nodes
}

// 判断是否结束
function isEnd(context: ParseContext, ancestor: TagElementType[]) {
  const s = context.source
  // console.log('isEnd --- source', s)
  // console.log('isEnd --- ancestor', ancestor)
  // 遇到结束标签时，例如：</div>
  if (startsWith(s, '</')) {
    for (let i = ancestor.length - 1; i >= 0; --i) {
      if (startsWithEndTagOpen(s, ancestor[i].tag)) {
        return true
      }
    }
  }
  // source 没有值时
  return !s
}

// 解析插值表达式
function parseInterpolation(context: ParseContext) {
  // {{message}} -> message

  const openDelimiter = '{{'
  const openDelimiterLength = openDelimiter.length
  const closeDelimiter = '}}'
  const closeDelimiterLength = closeDelimiter.length
  const closeIndex = context.source.indexOf(closeDelimiter, openDelimiterLength)
  // console.log('closeIndex', closeIndex)

  advanceBy(context, openDelimiterLength)
  // console.log('context.source', context.source)

  const rawContentLength = closeIndex - openDelimiterLength
  const rawContent = parseTextData(context, rawContentLength)
  const content = rawContent.trim()
  // console.log('content', content)

  advanceBy(context, closeDelimiterLength)
  // console.log('context.source', context.source)

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content
    }
  }
}

// 解析 Element
function parseElement(context: ParseContext, ancestor: TagElementType[]) {
  // 1. 解析 Tag
  const element: TagElementType =
    parseTag(context, TagTypes.START) || ({} as any)
  // 收集已解析到的 element 标签
  ancestor.push(element)

  element.children = parseChildren(context, ancestor)
  // 移除对应的 element 标签
  ancestor.pop()

  console.log('element ---- ', element)
  if (startsWithEndTagOpen(context.source, element.tag)) {
    parseTag(context, TagTypes.END)
  } else {
    throw new Error(`缺少结束标签：${element.tag}`)
  }

  return element
}
function parseTag(context: ParseContext, type: TagTypes) {
  // 1. 解析 Tag
  const match: any = /^<\/?([a-z]*)/i.exec(context.source)
  console.log('match', match)
  const tag = match[1]
  console.log('tag', tag)
  // 2. 删除处理完的代码
  advanceBy(context, match[0].length)
  advanceBy(context, 1)
  console.log('context', context)

  if (type === TagTypes.END) {
    return
  }
  return {
    type: NodeTypes.ELEMENT,
    tag
  }
}

// 默认解析 Text
function parseText(context: ParseContext) {
  let endIndex = context.source.length
  const endTokens = ['<', '{{']

  for (let i = 0; i < endTokens.length; i++) {
    const index = context.source.indexOf(endTokens[i])
    if (index !== -1 && endIndex > index) {
      endIndex = index
    }
  }

  const content = parseTextData(context, endIndex)

  console.log('parseText --- context.source', context.source)

  return { type: NodeTypes.TEXT, content }
}

function parseTextData(context: ParseContext, length: number) {
  const rawText = context.source.slice(0, length)
  advanceBy(context, length)
  return rawText
}

function startsWithEndTagOpen(source: string, tag: string) {
  return (
    startsWith(source, '</') &&
    source.slice(2, 2 + tag.length).toLocaleLowerCase() ===
      tag.toLocaleLowerCase()
  )
}

function advanceBy(context: ParseContext, length: number) {
  context.source = context.source.slice(length)
}

function createRoot(children) {
  return {
    type: NodeTypes.ROOT,
    children
  }
}

function createParseContext(content: string): ParseContext {
  return {
    source: content
  }
}
