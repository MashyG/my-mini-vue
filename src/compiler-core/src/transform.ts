import { NodeTypes } from './ast'
import { TO_DISPLAY_STRING } from './runtimeHelpers'

interface NodeTransforms {
  nodeTransforms: any[]
}
interface TransformContext {
  root: any
  nodeTransforms: NodeTransforms[]
  helpers: any
  helper: any
}

export function transform(root: any, options: NodeTransforms | any = {}) {
  const context = createTransformContext(root, options)
  // 1. 深度优先遍历
  traverseNode(root, context)

  // root.codegenNode
  createRootCodegen(root)

  root.helpers = [...context.helpers.keys()]
}

function createRootCodegen(root: any) {
  root.codegenNode = root.children[0]
}

function createTransformContext(
  root: any,
  options: NodeTransforms
): TransformContext {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    helper(key) {
      context.helpers.set(key, 1)
    }
  }
  return context
}

function traverseNode(node: any, context: TransformContext) {
  const { nodeTransforms } = context || {}
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transformFunc: any = nodeTransforms[i]
    transformFunc(node)
  }
  console.log('node', node)

  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      context.helper(TO_DISPLAY_STRING)
      break
    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      traverseChildren(node.children, context)
      break
    default:
      break
  }
}

function traverseChildren(children: any, context: TransformContext) {
  if (children) {
    for (let j = 0; j < children.length; j++) {
      const childNode = children[j]
      traverseNode(childNode, context)
    }
  }
}
