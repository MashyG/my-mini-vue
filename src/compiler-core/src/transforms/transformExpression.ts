import { NodeTypes } from '../ast'

export function transformExpression(node) {
  if (NodeTypes.INTERPOLATION) {
    node.content = processExpression(node.content)
  }
}

function processExpression(node: any) {
  node.content = `_ctx.${node.content}`
  return node
}
