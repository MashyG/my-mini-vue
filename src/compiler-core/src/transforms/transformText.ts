import { NodeTypes } from '../ast'
import { isText } from '../utils'

export function transformText(node: any) {
  if (node.type === NodeTypes.ELEMENT) {
    return () => {
      const children = node.children
      const { length } = children || []
      let currentContainer

      for (let i = 0; i < length; i++) {
        const child = children[i]
        // 看看下一个节点是不是 text 类
        if (isText(child)) {
          for (let j = i + 1; j < length; j++) {
            const nextChild = children[j]
            if (isText(nextChild)) {
              if (!currentContainer) {
                currentContainer = children[i] = {
                  type: NodeTypes.COMPOUND_EXPRESSION,
                  children: [child]
                }
              }

              currentContainer.children.push(' + ', nextChild)
              children.splice(j, 1)
              j--
            } else {
              currentContainer = undefined
              break
            }
          }
        }
      }
    }
  }
}
