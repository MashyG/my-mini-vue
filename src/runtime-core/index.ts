export { h } from './h'
export { renderSlots } from './helpers/renderSlots'
export { provide, inject } from './helpers/apiInject'
export { createTextVNode } from './vnode'
export { getCurrentInstance } from './component'
export { createRenderer } from './renderer'
