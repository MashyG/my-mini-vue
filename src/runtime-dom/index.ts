import { createRenderer } from '../runtime-core'

// createElement
function createElement(type) {
  console.log('createElement ----->>>>')
  return document.createElement(type)
}
// patchProps
function patchProps(el, key, prevVal, nextVal) {
  console.log('patchProps ----->>>>')
  const isEvent = (k: string) => /^on[A-Z]/.test(k)
  if (isEvent(key)) {
    const event = key.slice(2).toLocaleLowerCase()
    el.addEventListener(event, nextVal)
  } else {
    if (nextVal === undefined || nextVal === null) {
      el.removeAttribute(key)
    } else {
      el.setAttribute(key, nextVal)
    }
  }
}
// insert
function insert(el, parent) {
  console.log('insert ----->>>>')
  parent.append(el)
}

const renderer: any = createRenderer({
  createElement,
  patchProps,
  insert
})

export function createApp(...args) {
  return renderer.createApp(...args)
}

export * from '../runtime-core'
