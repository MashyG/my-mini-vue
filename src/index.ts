// mini-vue 出口
export * from './runtime-dom'

import { baseCompile } from './compiler-core/src'
import * as runtimeDom from './runtime-dom'

function complier2Function(template) {
  const { code } = baseCompile(template)
  const render = new Function('Vue', code)(runtimeDom)
  return render
}

runtimeDom.registerRuntimeCompiler(complier2Function)
