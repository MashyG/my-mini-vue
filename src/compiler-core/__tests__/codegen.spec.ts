import { generate } from '../src/codegen'
import { baseParse } from '../src/parse'
import { transform } from '../src/transform'
import { transformExpression } from '../src/transforms/transformExpression'

describe('codegen 实现 render', () => {
  // 利用 jest 的快照测试 -> toMatchSnapshot
  it('string 纯字符串', () => {
    const ast = baseParse('mashy')
    transform(ast)
    const { code } = generate(ast)

    expect(code).toMatchSnapshot()
  })

  it('interpolation 插值类型', () => {
    const ast = baseParse('{{mashy}}')
    transform(ast, {
      nodeTransforms: [transformExpression]
    })
    const { code } = generate(ast)

    expect(code).toMatchSnapshot()
  })
})
