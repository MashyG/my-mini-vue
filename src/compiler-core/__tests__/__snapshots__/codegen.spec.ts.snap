// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen 实现 render element 标签 1`] = `
"const { createElementVNode: _createElementVNode, toDisplayString: _toDisplayString } = Vue
return function render(_ctx, _cache) {
return _createElementVNode('div', null, 'hi~'_toDisplayString(mashy))
}"
`;

exports[`codegen 实现 render interpolation 插值类型 1`] = `
"const { toDisplayString: _toDisplayString } = Vue
return function render(_ctx, _cache) {
return _toDisplayString(mashy)
}"
`;

exports[`codegen 实现 render string 纯字符串 1`] = `
"return function render(_ctx, _cache) {
return 'mashy'
}"
`;
